File: decision_tree/c/dt.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Structure for storing the dataset
struct Dataset {
    double** data;
    int* target;
    int num_samples;
    int num_features;
};

// Function to load the dataset (Iris dataset as an example)
struct Dataset load_iris() {
    struct Dataset iris;
    // Load the dataset here...
    // Assign values to iris.data, iris.target, iris.num_samples, and iris.num_features
    return iris;
}

// Function to split the dataset into training and testing sets
void train_test_split(struct Dataset dataset, double test_size, int random_state,
                      double*** X_train, double*** X_test, int** y_train, int** y_test) {
    // Split the dataset here...
    // Assign values to *X_train, *X_test, *y_train, and *y_test
}

// Structure for the decision tree classifier
struct DecisionTreeClassifier {
    // Define the structure of the decision tree classifier here...
};

// Function to create a decision tree classifier
struct DecisionTreeClassifier create_classifier() {
    struct DecisionTreeClassifier classifier;
    // Create the decision tree classifier here...
    return classifier;
}

// Function to train the classifier on the training data
void train_classifier(struct DecisionTreeClassifier* classifier, double** X_train, int* y_train) {
    // Train the classifier on the training data here...
}

// Function to make predictions on the test data
void predict(struct DecisionTreeClassifier classifier, double** X_test, int* y_pred, int num_samples) {
    // Make predictions on the test data here...
    // Assign values to y_pred
}

// Function to calculate the accuracy of the model
double accuracy_score(int* y_test, int* y_pred, int num_samples) {
    // Calculate the accuracy of the model here...
    // Return the accuracy value
}

int main() {
    // Load the dataset
    struct Dataset iris = load_iris();

    // Split the dataset into training and testing sets
    double** X_train;
    double** X_test;
    int* y_train;
    int* y_test;
    train_test_split(iris, 0.2, 42, &X_train, &X_test, &y_train, &y_test);

    // Create a decision tree classifier
    struct DecisionTreeClassifier classifier = create_classifier();

    // Train the classifier on the training data
    train_classifier(&classifier, X_train, y_train);

    // Make predictions on the test data
    int* y_pred = (int*)malloc(iris.num_samples * sizeof(int));
    predict(classifier, X_test, y_pred, iris.num_samples);

    // Calculate the accuracy of the model
    double accuracy = accuracy_score(y_test, y_pred, iris.num_samples);
    printf("Accuracy: %f\n", accuracy);

    // Free dynamically allocated memory
    // ...

    return 0;
}

File: decision_tree/zig/dt.zig
const std = @import("std");
const csv = std.encoding.csv;
const DecisionTree = @import("decision_tree");

pub fn main() !void {
    // Load the dataset (Iris dataset as an example)
    const irisData = try std.fs.cwd().openFile("iris.csv", .{ .read = true });
    defer irisData.close();

    var csvReader = csv.Reader.init(irisData.reader());
    const header = try csvReader.next();
    var irisRows = []std.ArrayList(u8).init(std.heap.page_allocator);

    while (try csvReader.next()) |row| {
        const sepalLength = std.fmt.parseFloat(row[0]) catch unreachable;
        const sepalWidth = std.fmt.parseFloat(row[1]) catch unreachable;
        const petalLength = std.fmt.parseFloat(row[2]) catch unreachable;
        const petalWidth = std.fmt.parseFloat(row[3]) catch unreachable;
        const species = std.fmt.parseEnum("Species", row[4]) catch unreachable;

        var rowData = std.ArrayList(u8).init(std.heap.page_allocator);
        try rowData.appendAll(u8, &[_]u8{ sepalLength, sepalWidth, petalLength, petalWidth });
        irisRows.append(rowData);
    }

    const splitRatio: f32 = 0.8;
    const splitIndex = @divTrunc(irisRows.len, 1.0 / splitRatio);
    const trainingData = irisRows[0..splitIndex];
    const testingData = irisRows[splitIndex..];

    // Create a decision tree classifier
    var tree = DecisionTree.ID3Tree.init(4);

    // Train the classifier on the training data
    for (trainingData) |rowData, _index| {
        const features = rowData.items;
        const label = std.fmt.parseEnum(u8, "Species", header.items[4]) catch unreachable;
        try tree.train(features, label);
    }

    // Make predictions on the test data
    var predictions = []std.ArrayList(u8).init(std.heap.page_allocator);
    for (testingData) |rowData, _index| {
        const features = rowData.items;
        const prediction = try tree.predict(features);
        predictions.append(prediction);
    }

    // Calculate the accuracy of the model
    const accuracy = accuracyScore(predictions, testingData);
    std.log.info("Accuracy: {}", .{accuracy});
}

fn accuracyScore(predictions: []std.ArrayList(u8), testData: []std.ArrayList(u8)) f32 {
    var correctPredictions: u32 = 0;
    const n = predictions.len;
    for (predictions) |prediction, index| {
        const actual = testData[index].items[4];
        if (prediction.items[0] == actual) {
            correctPredictions += 1;
        }
    }

    return f32(correctPredictions) / f32(n);
}
